[libc] Use best-fit binary trie to make malloc logarithmic

mysterymath opened the pull request:
This reworks the free store implementation in libc's malloc to use a dlmalloc-style binary trie of circularly linked FIFO free lists. This data structure can be maintained in logarithmic time, but it still permits a relatively small implementation compared to other logarithmic-time ordered maps.

The implementation doesn't do the various bitwise tricks or optimizations used in actual dlmalloc; it instead optimizes for (relative) readability and minimum code size. Specific optimization can be added as necessary given future profiling.

mysterymath committed:
[libc] Use best-fit binary trie to make malloc logarithmic

This reworks the free store implementation in libc's malloc to use a
dlmalloc-style binary trie of circularly linked FIFO free lists. This
data structure can be maintained in logarithmic time, but it still
permits a relatively small implementation compared to other
logarithmic-time ordered maps.

The implementation doesn't do the various bitwise tricks or
optimizations used in actual dlmalloc; it instead optimizes for
(relative) readability and minimum code size. Specific optimization can
be added as necessary given future profiling.

mysterymath commented:
@nopsledder

michaelrj-google commented on libc/src/__support/freelist_heap.h:
```diff
@@ -54,89 +47,75 @@ template <size_t NUM_BUCKETS = DEFAULT_BUCKETS.size()> class FreeListHeap {
-  BlockType *block = *result;
-  freelist_.add_chunk(block_to_span(block));
-  is_initialized_ = true;
+inline void FreeListHeap::init() {
```
`inline` -> `LIBC_INLINE` in multiple places

mysterymath replied:
Done.

michaelrj-google commented on libc/src/__support/freetrie.h:
```diff
@@ -0,0 +1,253 @@
+//===-- Interface for freetrie
```
nit: formatting

mysterymath replied:
Done.

michaelrj-google commented on libc/test/src/__support/freelist_test.cpp:
```diff
@@ -8,159 +8,43 @@
-using LIBC_NAMESPACE::cpp::array;
-using LIBC_NAMESPACE::cpp::byte;
-using LIBC_NAMESPACE::cpp::span;
+namespace LIBC_NAMESPACE_DECL {
```
In General we try to avoid putting the tests inside of the `LIBC_NAMESPACE` to ensure that we're getting our implementation of a function instead of one pulled from the system. I think it's probably fine to do `using` here since it's only testing internals, but I'd like to avoid the `namespace LIBC_NAMESPACE_DECL`.

mysterymath replied:
Done.

michaelrj-google reviewed (commented):
Looks fine overall, I haven't done a deep review of the logic but I did comment on a few style things.

mysterymath committed:
Address michaelrj-google review comments

nopsledder reviewed (commented):
Thanks for this! Most of my comments are high level; let me know if you want to discuss more.

mysterymath committed:
Address nopsledder review comments

mysterymath committed:
Add explanation for clearing deferred subtrie

mysterymath committed:
Add docs; minor condition change to ensure logarithmic time in corner case

mysterymath committed:
Format

mysterymath committed:
Revert linux entrypoint edit used for testing (whoops)

mysterymath committed:
Add a fuzzing test

mysterymath committed:
Fix issues found by fuzzer

These were found almost immediately, and some of them were quite tricky.
Hooray for fuzzing!

mysterymath commented:
I wasn't sufficiently confident that this worked correctly, so I wrote a fuzzer test. Thankfully, this found quite a few issues, so I've fixed those and included the fuzzer. Hooray for property based testing!

michaelrj-google commented on libc/src/__support/freelist.h:
```diff
@@ -9,200 +9,80 @@
+  /// Push a block to the back of the list.
+  /// The block must be large enough to contain a node.
+  LIBC_INLINE void push(Block<> *block) {
+    LIBC_ASSERT(!block->used() &&
+                "only free blocks can be placed on free lists");
+    LIBC_ASSERT(block->inner_size_free() >= sizeof(FreeList) &&
+                "block too small to accomodate free list node");
```
Would making this fail in non-debug mode be useful? This seems like the sort of thing that would be useful to catch at runtime.

michaelrj-google commented on libc/src/__support/freelist_heap.h:
```diff
@@ -28,23 +28,16 @@ extern "C" cpp::byte __llvm_libc_heap_limit;
-                                                       128, 256, 512};
-
-template <size_t NUM_BUCKETS = DEFAULT_BUCKETS.size()> class FreeListHeap {
+class FreeListHeap {
```
I like `FreelistHeapAlloc` or similar to make it clear this is an allocator. "Heap" is what we call the memory we allocate outside of the stack, but it's also the name of a generic data structure.

mysterymath replied:
From the above, I like `EmbeddedHeapAlloc`; either should fit the intended use case well. But if it's alright, I think I'd like to do a NFC reorganization PR afterwards; this is already a pretty enormous change.

michaelrj-google commented on libc/src/__support/freetrie.h:
```diff
@@ -0,0 +1,237 @@
+    /// @returns The lower half of the size range.
+    LIBC_INLINE SizeRange lower() const { return {min, width / 2}; }
+
+    /// @returns The lower half of the size range.
```
nit: fix this comment

mysterymath replied:
Done, thanks.

michaelrj-google reviewed (commented):
Some minor comments, I didn't notice anything wrong with it but I'm not super confident that I understood it fully.

mysterymath committed:
Remove `using namespace LIBC_NAMESPACE;` from fuzzer

mysterymath committed:
Fix comment

mysterymath committed:
clang-format

mysterymath commented:
AFAICT the buildkite issue is a flake with the new `generate_test_report.py`; I've reported this in https://github.com/llvm/llvm-project/pull/113447.

mysterymath commented:
The CI flake has been fixed, and it looks green. Friendly ping; anything left to do on this one?

PiJoules reviewed (approved):
LGTM Thanks!
